### PR TITLE
Delay launch of pure noncombatant fighters & drones

### PR DESCRIPTION
**Pull request by ravenshining**
## Summary
This PR adds some logic to Command::DEPLOY and AI::ShouldDock in AI.cpp for unarmed carried ships.  If a fighter or drone has no usable weapons, no anti-missile turrets, and no scanners (e.g., a Boxwing or a Faes'mar), it will now refrain from launching when the Deploy command is issued until its parent ship is below 3/8 health (from full with no disruption, 75% hull).

Once launched, said unarmed ships - and pure missile fighters that have run out of ammo - will attempt to return to their parent if either the parent is above 50% health, or they are at lower health than their parent.

This will allow a carrier with a mix of cargo drones and combat fighters to safely launch its fighters without worrying about losing its cargo drones, unless the carrier itself is in danger and wishes to shed the weight of and shield itself with its cargo drones.

The exemption from this behaviour for fighters/drones with scanners preserves the behaviour of Navy Surveillance Cruisers. It has a side effect of allowing the player to specify they want a particular fighter to always launch by placing a cargo scanner on it.

The object of this PR was to open up possibilities for more noncombatant carried ships without them acting silly, and to open up possibilities for practical mixed boxwing-fighter or missile fighter use.

## Save File
This save file has a player with five Tek Far 71 Lek's, five Skeins, and a diverse complement of fighters and drones to test.  The player is in the Tarazed system where fighters and carriers may be purchased, and has been around long enough for the FW carriers to start appearing.  It was created with the All-Content plugin, which is not necessary to have installed.

[Smart Boxwings.txt](https://github.com/endless-sky/endless-sky/files/4500619/Smart.Boxwings.txt)

## Testing Done
Flying around with the above save, fighting and watching how my ships react, firing on my own ships to verify launch behaviour, spawning in AI-controlled ships with Boxwings to see how they react, observing Navy Surveillance cruisers to make sure they're unaffected.

Compiled & tested on a Debian 8 machine.

-----------------------

## UI Screenshots

Showing fleet after a Deploy command issued, note the Faes'mar have not launched:

no-launched-boxwings

After some combat, a Seli'maar equipped with Sidewinders has run out of ammo and returned to its mothership, while another that still has ammo remains in space.

out-of-ammo

While firing on this escort, it launches its Faes'maars having lost sheilds.

launched-boxwings1

## Performance Impact
Possible negative impact as it is adding complexity to the AI code.

### CODE REVIEW

File: source/AI.cpp [hunk 1]
```diff
@@ -116,10 +116,31 @@ namespace {
...
+			return true;
+		return false;
+	}
+	
+	// Launch fighters and drones.
 	void Deploy(const Ship &ship, bool includingDamaged)
 	{
 		for(const Ship::Bay &bay : ship.Bays())
-			if(bay.ship && (includingDamaged || bay.ship->Health() > .75))
+			if(bay.ship && (includingDamaged || ( bay.ship->Health() > .75 && ( IsArmed(*bay.ship) || CanAntimissile(*bay.ship) || CanScan(*bay.ship) || (ship.Shields() + ship.Hull() < 0.75 )))))
```
[the remark this answers is not in the document]
Comment (ravenshining) [reply source/AI.cpp:143]: I tried using Health() first, but it seemed the Deploy and Return had different ideas of what Health() meant - If I set the threshold to 0.5, for example, they would deploy at 0.5 (sheilds+hull), but not return unless hull was over 0.5 - which meant never if the test ship did not have hull repair.  I saw elsewhere Sheilds() and Hull() being used - (at line 602 for appeasement, 1144 for preferential targeting, & 2786 for Frugal behaviour) - and copied the convention.

**Comment by ravenshining** [comment]
Thank you for the review!  I'll look into implementing your suggestions this weekend.

**Comment by Amazinite** [comment]
> How much of this is a non-issue when we support custom launch groups for carried ships?

Ref: #868

**Comment by ravenshining** [comment]
Regarding #868, yes, a separate keybind might be preferable for player ships - or might not, depending on how much one likes to micromanage.  This PR started off as a way to govern AI behaviour, and the effect on player ships is more of a side-benefit.  I could add - in addition to the personality checks tehhowich requested - an !ship.IsYours() in here if we don't want this automatic behaviour to affect player ships and have new keybinds instead.

**Comment by Amazinite** [comment]
> Regarding #868, yes, a separate keybind might be preferable for player ships - or might not, depending on how much one likes to micromanage.

It's already possible to create escort groups bound to the number keys, and those groups will follow the orders you give them (gather, hold, attack, move) without that order getting issued to those escorts outside the group, but deploy is not one such command that only gets issued to the selected fighters; the deploy is currently fleet-wide.

**Comment by ravenshining** [comment]
Oh, I didn't realise that was a thing.

How did I not see all these other PRs and issues when researching prior to writing this?  They never came up in my searches...

Well, anyway, those still just apply to the player... except the ammo replenishment ones.  My main goal, having some sane behaviour for AI ships with unarmed fighters, remains outside the scope of those.

**Comment by tehhowch** [comment]
#4719 was merged, which resolves #868, which I believe means the player has full control over the deployment state of noncombatant fighters & drones. Closing this since I don't think there is anything this PR offers that is not offered elsewhere except custom behavior to NPCs, which can already be designed to not suffer from this issue.

**Pull request closed without merging**